Provide a synchronous API for PayIDClient

keefertaylor opened the pull request:
## High Level Overview of Change

Add threading logic for `PayIDClient`.

### Context of Change

This change:
- Runs all networking on a dedicated queue
- Allows users to specify a callback queue (defaults to the main queue, which is the same behavior as before)
- Provides a synchronous method to fetch PayIDs
- Updates a test to prove the synchronous method works. 

This is largely motivated by it being [hard to write demo code](https://github.com/xpring-eng/Xpring-SDK-Demo/pull/39) but likely just makes developer's lives easier overall.

### Type of Change

<!--
Please check relevant options, delete irrelevant ones.
-->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Tests (You added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation Updates
- [ ] Release

## Before / After

New APIs introduced to make working with PayID easier.

## Test Plan

CI 

## Future Tasks
We'll need to do something similar in `XRPPayIDClient` and `XpringClient`.

Shmaff commented on XpringKit/PayID/PayIDClient.swift:
```diff
@@ -29,20 +33,51 @@ public class PayIDClient {
   ) {
+    // Wrap completion calls in a closure which will dispatch to the callback queue.
+    let queueSafeCompletion: (Result<CryptoAddressDetails, PayIDError>) -> Void = {
+      completion($0)
```
Missing use of the callbackQueue provided as a parameter to this address function. As it is now, it will always return on the calling thread.

keefertaylor commented:
Fixed queue callbacks. Thank you for the catch. 

I added a test to prove this works. Unfortunately, it's kind of hard to test this logic, so the best I can do is just say "we call back on something other than the main thread". I'm very open to pointers if there's a better test I can write here. 

Testing for PayID is a bit touch and go because we had a bunch of issues with the OpenAPI generating testable code, so unit tests are still outstanding and we do all of our testing in integration tests. #238 addresses this so things will get better soon.

intelliot commented on XpringKit/PayID/PayIDClient.swift:
```diff
@@ -29,20 +33,53 @@ public class PayIDClient {
   ///
   //  TODO: Link a canonical list at payid.org when available.
-  public init(network: String) {
+  public init(network: String, completionQueue: DispatchQueue = DispatchQueue.main) {
```
is this needed?

Shmaff replied:
Agreed. This can be removed. The network and callback queues are already in place. Plus, AFAIK, this is a breaking change as far as bitcode is concerned since it's a public API.

keefertaylor replied:
Apologies. I was heads down on the PayID release last week. Sorry to be sloppy and send out incomplete PRs. 

This is resolved.

Shmaff commented:
@keefertaylor I did a little investigation as to what's possible for testing in terms of what queue a thread is currently executing on.

I used this (https://stackoverflow.com/a/60314121) to test locally with success for all queue types (that I tested: main, global() and custom). One step I missed initially was to `registerDetection` of a custom queue.

keefertaylor commented on XpringKit/PayID/PayIDClient.swift:
```diff
@@ -17,9 +17,13 @@ public class PayIDClient {
   /// Initialize a new PayID client.
   ///
-  /// - Parameter network: The network that addresses will be resolved on.
+  /// - Parameters:
```
Removed.

keefertaylor commented:
@Shmaff It worries me a little bit that Apple seems to hide this functionality (which makes me think it is on purpose or it has more edge cases than we think). That said, I can't make this code break so I pulled it in. 

PTAL

Shmaff commented:
@keefertaylor Good point. However, Apple isn't known for their documentation 😉

Shmaff commented on Tests/PayID/PayIDIntegrationTests.swift:
```diff
@@ -32,6 +33,26 @@ final class PayIDIntegrationTests: XCTestCase {
+
+    // WHEN it is resolved to an XRP address and provided a custom queue and not on the main thread.
+    payIDClient.address(for: .testPointer, callbackQueue: customCallbackQueue) { _ in
+      XCTAssertEqual(DispatchQueue.currentQueueLabel, queueLabel)
```
👍🏼
